src/snowflake/snowpark/_internal/compiler/repeated_subquery_elimination.py

```diff
@@ -171,6 +171,7 @@ def _update_parents(
                     with_block._is_valid_for_replacement = True
 
                     resolved_with_block = self._query_generator.resolve(with_block)
+                    resolved_with_block._is_valid_for_replacement = True
                     resolved_with_block_map[
```

Comment:
we set this flag here https://github.com/snowflakedb/snowpark-python/blob/main/src/snowflake/snowpark/_internal/compiler/query_generator.py#L225, do we have a case that didn't go into the do_resolve_with_resolved_children step, maybe we should overwrite at the do_resolve step also

Reply:
What was the consequence of missing it? a missed optimization opportunity or a correctness issue?
Do we need to add more test coverage for the cte optimization?

Reply:
the optimization may fail. @sfc-gh-aalam maybe it would be good to add a try catch for the optimization stage, and fallback to the original query if the optimization fails.

Reply:
https://snowflakecomputing.atlassian.net/browse/SNOW-1844094

Reply:
<img width="2215" alt="Screenshot 2024-12-03 at 2 43 49 PM" src="https://github.com/user-attachments/assets/5123ef47-71c3-4387-86dc-cedc2e7250a6">

I'm investigating the source of this issue. But the problem is not with the `SnowflakePlan` but the `SelectSnowflakePlan` that is generated afterwards.